include/net/sock.h

```diff
@@ -507,16 +507,13 @@ struct sock {
 	void			(*sk_data_ready)(struct sock *sk);
 	void			(*sk_write_space)(struct sock *sk);
 #ifdef CONFIG_SECURITY_TEMPESTA
-	int			(*sk_prepare_xmit)(struct sock *sk,
-						   struct sk_buff *skb,
-						   unsigned int mss_now,
-						   unsigned int *limit,
-						   unsigned int *skbs);
 	int			(*sk_write_xmit)(struct sock *sk,
 						 struct sk_buff *skb,
 						 unsigned int mss_now,
-						 unsigned int limit,
-						 unsigned int skbs);
+						 unsigned int limit);
+	int			(*sk_fill_write_queue)(struct sock *sk,
```

Comment:
I think we should document this breaking change, explaining why `sk_prepare_xmit` is deprecated, and why the new `sk_fill_write_queue` requires some somewhat complicated limit calculations. If not, maybe the author himself will forget about it after a few weeks. It's better to document it in the code as a comment, because https://github.com/tempesta-tech/tempesta/issues/1196 only discusses some theoretical design (which is even outdated after the code rewrite), not the actual implementation.

Reply:
Yeah, it could make sense to document both the two callback, when they are called and what they do

Reply:
Personally, I don't quite understand why we moved frames from "prepare_xmit" to "tcp_push_pending_frames", which is the biggest change in this PR. In `prepare_xmit`, we already know that TCP has enough window to send data, but in `tcp_push_pending_frames`, we have to calculate it ourselves, which seems to be duplication of work. Could you clarify this further in comments?

Reply:
Yes, I will add comment before merge PR. There were several problems with the previous approach:
1. We push response skb directly to socket write queue. If skb belongs to stream, which is blocked by exceeding HTTP2 stream window, we can't  send this skb to the client. This skb blockes all other skb in socket write queue, until window update will come.
2. We can't prioritize streams. We push responses to socket write queue directly, so there is no prioritization. Now we push response skbs to our private scheduler queue. Then we call `tcp_push_pending_frames` where we run our scheduler to choose the most priority stream and send data according TCP window. Also `tcp_push_pending_frames` called from `tcp_data_snd_check` when we receive data from the peer and TCP window opened again.